### ISSUE TITLE
uiLinkLabel Control

### ISSUE THREAD

**Issue by tom-corwin** [andlabs/libui#363]
Currently there is no support for this, but it would be beneficial as a basic control. It would allow to get and set the text, as well as support an event which would happen when the user clicks the uiLinkLabel.

Some form of public API could look like this:

```c
typedef struct uiLinkLabel uiLinkLabel;
#define uiLinkLabel(this) ((uiLinkLabel *) (this))
_UI_EXTERN char *uiLinkLabelText(uiLinkLabel *l);
_UI_EXTERN void uiLinkLabelSetText(uiLinkLabel *l, const char *text);
_UI_EXTERN void uiLinkLabelOnClicked(uiLinkLabel *l, void (*f)(uiLinkLabel *l, void *data), void *data);
_UI_EXTERN uiLinkLabel *uiNewLinkLabel(const char *text);
```

This would allow for the developer to have a choice of a uiButton, or a uiLinkLabel to convey text that is linked to a clickable event.

I know that there is a LinkLabel control on Windows, but I'm not sure how this would be done using Cocoa or GTK.

**Comment by tom-corwin** [comment]
Also, the text displayed by a uiLinkLabel should be formatted as if it were a hyperlink, but when clicked would fire the uiLinkLabelOnClicked event.

**Comment by mischnic** [comment]
With this design, it wouldn't be possible to have a link in a running text.

Related previous issues: #303, #52

**Comment by cody271** [comment]
So this would be a more special purpose control than a flat button? Would the text have to be a valid URL then?

API wise, I think this could just be a "subclass" of `uiButton`:
```c
_UI_EXTERN uiButton *uiNewLinkButton(const char *text);
```

**Comment by tom-corwin** [comment]
Rather than *just supporting URLs*, this would allow for the user to display a label that when clicked, will call an event, which could do whatever you wanted.

And @cody271, I would consider it a uiLabel, since they're about the same, except it's text is colored, underlined, and clickable.

**Comment by mischnic** [comment]
> display a label that when clicked, will call an event, which could do whatever you wanted.

So basically a button which looks like a label.

**Comment by cody271** [comment]
Yes, still voting for `uiButton` subclass here. That's how it would be implemented, and all the `uiButton` methods would still apply..

**Comment by tom-corwin** [comment]
Point taken, you're both right.

**Comment by andlabs** [comment]
Richer label content is an active problem I am thinking about. The easiest solution would indeed be a link-only uiButton-like control; that was my original plan in the meantime.

**Comment by cody271** [comment]
The use of rich label content needs to be explicit here..
I'm thinking we add a flat button subclass, and then another constructor for this link button.

Something like this:
```c
_UI_EXTERN uiButton *uiNewFlatButton(const char *text);
_UI_EXTERN uiButton *uiNewFlatButtonWithURL(const char *url);
```

**Comment by mischnic** [comment]
I don't think `uiNewFlatButtonWithURL` is ideal, because it should be possible to have a link with the text "Homepage" (and not "https://www.example.com/xyz").

The link button would interally call `uiOpenURL` (which should also be exposed) and that was previously discussed - essentially: what if the url is malformed (e.g. wrong protocol or no handler for protocol), the OSs aren't handling this consistently.

Related issues: #160  #52
